contributors/design-proposals/network/service-topology-key.md

```diff
@@ -0,0 +1,171 @@
+# topology aware routing of services
+
+Author: @m1093782566
+
+## Objective
+
+Figure out a generic way to implement the "local service" route, say "topology aware routing of service". 
+
+Locality is defined by user, it can be any topology-related thing. "Local" means the "same topology level", e.g. same node, same rack, same failure zone, same failure region, same cloud provider etc.
+
+## GOAL
+
+A generic way to support various topology levels, regarless whatever the topology key is.
+
+## Non-goal
+
+Scheduler spreading to implement this sort of topology guarantee.
+
+## Use cases
+
+* Logging agents such as fluentd. Deploy fluentd as DaemonSet and applications only need to communicate with the fluentd in the same node.
+* For a sharded service that keeps per-node local information in each shard.
+* Authenticating proxies such as [aws-es-proxy](https://github.com/kopeio/aws-es-proxy).
+* In container identity wg, being able to give daemonset pods a unique identity per host is on the 2018 plan, and ensuring local pods can communicate to local node services securely is a key goal there. -- from @smarterclayton
+* Regional data costs in multi-AZ setup - for instance, in AWS, with a multi-AZ setup, half of the traffic will switch AZ, incurring regional data Transfer costs, whereas if something was local, it wouldn't hit the network.
```

Comment:
[Many](https://aws.amazon.com/blogs/compute/kubernetes-clusters-aws-kops/) [guides](https://www.nivenly.com/k8s-aws-private-networking/) tell you to use three AZs, in which case the number will be two-thirds not half.

Reply:
ok, :)